x/tally/keeper/filter.go

```diff
@@ -28,69 +71,42 @@ type FilterResult struct {
 // index i is an outlier, consensus boolean, consensus data proxy
 // public keys, and error. It assumes that the reveals and their
 // proxy public keys are sorted.
-func (k Keeper) ApplyFilter(ctx sdk.Context, input []byte, reveals []types.RevealBody, replicationFactor uint16) (FilterResult, error) {
+func ApplyFilter(filter types.Filter, reveals []types.RevealBody) (FilterResult, error) {
 	var result FilterResult
+	result.Errors = make([]bool, len(reveals))
 	result.Outliers = make([]int, len(reveals))
+	result.GasUsed = filter.GasCost()
 
-	if len(input) == 0 {
-		return result, types.ErrInvalidFilterType
-	}
-
-	// Determine basic consensus on tuple of (exit_code, proxy_pub_keys)
+	// Determine basic consensus on tuple of (exit_code_success, proxy_pub_keys)
 	var maxFreq int
-	var proxyPubKeys []string
 	freq := make(map[string]int, len(reveals))
-	for _, reveal := range reveals {
+	for i, reveal := range reveals {
 		success := reveal.ExitCode == 0
+		result.Errors[i] = !success
 		tuple := fmt.Sprintf("%v%v", success, reveal.ProxyPubKeys)
 		freq[tuple]++
 
 		if freq[tuple] > maxFreq {
-			proxyPubKeys = reveal.ProxyPubKeys
+			result.ProxyPubKeys = reveal.ProxyPubKeys
 			maxFreq = freq[tuple]
 		}
 	}
-	if maxFreq*3 < len(reveals)*2 {
+	if maxFreq*3 < len(reveals)*2 { // TODO remove basic consensus check?
 		return result, types.ErrNoBasicConsensus
 	}
-	result.ProxyPubKeys = proxyPubKeys
 
-	params, err := k.GetParams(ctx)
-	if err != nil {
-		return result, err
-	}
-
-	var filter types.Filter
-	switch input[0] {
-	case filterTypeNone:
-		filter, err = types.NewFilterNone(input)
-		result.GasUsed = params.FilterGasCostNone
-	case filterTypeMode:
-		filter, err = types.NewFilterMode(input)
-		result.GasUsed = params.FilterGasCostMultiplierMode * uint64(replicationFactor)
-	case filterTypeStdDev:
-		filter, err = types.NewFilterStdDev(input)
-		result.GasUsed = params.FilterGasCostMultiplierStddev * uint64(replicationFactor)
-	default:
-		return result, types.ErrInvalidFilterType
-	}
-	if err != nil {
-		result.GasUsed = 0
-		return result, err
-	}
+	outliers, consensus := filter.ApplyFilter(reveals, result.Errors)
 
-	outliers, err := filter.ApplyFilter(reveals)
 	switch {
-	case err == nil:
-		result.Outliers = outliers
+	case errorCount(result.Errors)*3 > len(reveals)*2:
+		result.Consensus = true
+		return result, types.ErrConsensusInError
```

Comment:
We could have outliers here right? And I think marking it as an error makes it more difficult to reason about at the callsite as this error and `ErrNoConsensus` will still have to go through phase III (determining gas used and payouts).

Reply:
My reasoning was that we return an error here if we do not want tally VM execution. 
I forgot that we need outlier list when we need to calculate the payouts. Added.